### PR TITLE
WSLC: Add UDP and IPv6 support for exposed image ports

### PR DESCRIPTION
**Pull request by dkbennett**
## Summary of the Pull Request

Support IPv6 and UDP in image-declared exposed ports. The `wslc -P`/`--publish-all` path now publishes every port from an image's `ExposedPorts` config on both IPv4 and IPv6 loopback, for both TCP and UDP — previously it silently published only TCP on IPv4 loopback. UDP publishing is conditional on networking relay (which is TCP only) and emits a warning and gracefully skips.

## PR Checklist

- [ ] **Closes:** Link to issue #xxx
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated if needed and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated if needed
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/wsl/) and link it here: #xxx

## Detailed Description of the Pull Request / Additional comments
When a container is created with `-P`/`--publish-all`, wslc reads the image's `ExposedPorts` config and publishes each declared port to an ephemeral host port. The prior implementation had two limitations:

- **UDP was dropped.** The loop skipped any non-TCP protocol (`if (protocol != IPPROTO_TCP) continue;`), so a `EXPOSE 53/udp` port was never published.
- **IPv6 was unsupported.** Each mapping hardcoded `AF_INET` and `127.0.0.1`, so exposed ports were reachable only over IPv4 loopback.

An image's `ExposedPorts` keys carry only a port and protocol (`port/tcp` or `port/udp`) and never an address family, so the code has to choose what to bind. This change mirrors Docker's documented `-P` default, which publishes to all host addresses on both IPv4 and IPv6, while preserving wslc's convention of binding loopback by default. Each exposed port is now published **dual-stack**: one mapping on `127.0.0.1` (IPv4) and one on `::1` (IPv6), preserving the port's declared protocol (tcp/udp).

### Networking-mode handling for UDP

wslc has two port-forwarding backends. The virtioNet path (`MapVirtioNetPort`, Consomme mode) maps ports at the netstack level and natively supports UDP and IPv6. The userspace wslrelay path (NAT mode, or Consomme with the wslrelay feature flag) is a TCP-only stream relay — `WSLCVirtualMachine::MapPort()` rejects any non-TCP mapping with `ERROR_NOT_SUPPORTED`.

Because publish-all previously skipped UDP unconditionally, an image exposing a UDP port worked (the port was just dropped) regardless of mode. Now that UDP mappings are generated, they would fail the entire container create/start under the wslrelay path. To avoid that regression, the exposed-ports loop now skips UDP ports when `UseWslRelayPortForwarding()` is active and emits a one-time user warning (`MessageWslcPublishAllUdpNotSupported`). TCP (IPv4 + IPv6) publishes normally in all modes; UDP publishes fully on the virtioNet path. This guard is self-retiring — it becomes a no-op automatically once the relay path is no longer used. Adding UDP support to the wslrelay stream relay is potential follow-up work; it is a substantial cross-boundary feature (datagram framing over hvsocket, per-source session tracking) and unnecessary given virtioNet already covers UDP.

### Testing

Added end-to-end test `WSLCE2E_Container_Create_PublishAll_DualStack` (`test/windows/wslc/e2e/WSLCE2EContainerCreateTests.cpp`). It builds an image with `EXPOSE 8080/tcp` and `EXPOSE 9090/udp`, runs `-P`, and asserts that each port is published with exactly two bindings — one on `127.0.0.1` and one on `::1` — each on a non-zero ephemeral host port. This covers the full `{tcp, udp} × {ipv4, ipv6}` matrix in a single test. Verified passing against a locally built and deployed x64 Debug build.

## Validation Steps Performed
* Newly added E2E test passes locally.

### CODE REVIEW

File: src/windows/wslcsession/WSLCContainer.cpp [hunk 1]
```diff
@@ -1824,18 +1824,18 @@ std::shared_ptr<WSLCContainerImpl> WSLCContainerImpl::Create(
             {
                 auto [port, protocol] = ParseExposedPortKey(portKey);
 
-                // Only TCP localhost mappings are currently supported by the relay path.
-                if (protocol != IPPROTO_TCP)
+                // Exposed ports carry only a port and protocol (tcp/udp), never an address family.
+                // Mirror Docker's dual-stack default by publishing each exposed port on both the IPv4
+                // and IPv6 loopback, while keeping wslc's loopback-only default binding convention.
+                for (const auto& [family, address] : {std::pair{AF_INET, "127.0.0.1"}, std::pair{AF_INET6, "::1"}})
                 {
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslcsession/WSLCContainer.cpp:1851]: Good catch here, the test passes due to running a mode that doesn't have the wslrelay feature flag. More work to be done here to make sure this path is also good.

[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslcsession/WSLCContainer.cpp:1851]: Added a warning if we are in this configuration and a graceful skip over UDP ports if we are using the relay, which is TCP-only.

Since we expect virtio / consomme to be supported in GA and that handles UDP ports already, this warning is just a graceful cover for the behavior in an edge case scenario.

File: test/windows/wslc/e2e/WSLCE2EContainerCreateTests.cpp [hunk 2]
```diff
@@ -1379,6 +1379,73 @@ class WSLCE2EContainerCreateTests
...
+        result.Verify({.Stderr = L"", .ExitCode = 0});
+
+        result = RunWslc(std::format(L"container start {}", WslcContainerName));
+        result.Verify({.Stderr = L"", .ExitCode = 0});
+
+        // Each exposed port (tcp and udp) must be published dual-stack: one IPv4 loopback and one IPv6
+        // loopback binding, each on an ephemeral host port.
+        const auto inspect = InspectContainer(WslcContainerName);
+        for (const auto& portKey : {std::string{"8080/tcp"}, std::string{"9090/udp"}})
+        {
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply test/windows/wslc/e2e/WSLCE2EContainerCreateTests.cpp]: Test only ever runs on the default virtioNet session (relay is opt-in), so it's deterministic. I am keeping it strict on purpose so it fails loudly if the default config changes rather than masking it and hiding potential other issues.

File: src/windows/wslcsession/WSLCContainer.cpp [hunk 3]
```diff
@@ -1820,22 +1820,39 @@ std::shared_ptr<WSLCContainerImpl> WSLCContainerImpl::Create(
...
+                // Mirror Docker's dual-stack default by publishing each exposed port on both the IPv4
+                // and IPv6 loopback, while keeping wslc's loopback-only default binding convention.
+                for (const auto& [family, address] : {std::pair{AF_INET, "127.0.0.1"}, std::pair{AF_INET6, "::1"}})
+                {
+                    auto& createdPort = ports.emplace_back();
+                    createdPort.HostPort = WSLC_EPHEMERAL_PORT;
+                    createdPort.Family = family;
+                    createdPort.ContainerPort = port;
+                    createdPort.Protocol = protocol;
+                    strcpy_s(createdPort.BindingAddress, address);
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslcsession/WSLCContainer.cpp:1857]: Good catch on the port-0 issue, but the suggested loop-only fix was too narrow: the same bug hits explicit `-p <ctr>` ephemeral ports, which go through `BuildPortMappings`, not the publish-all loop. Fixed at the root instead -- `BuildPortMappings` now pre-allocates a concrete host port when `UseWslRelayPortForwarding()` is true (it was checking only `NetworkingMode() == NAT`, a stale condition predating the Consomme + relay flag). One place now covers both publish-all and explicit ephemeral ports.

File: src/windows/wslcsession/WSLCContainer.cpp [hunk 4]
```diff
@@ -1820,22 +1823,39 @@ std::shared_ptr<WSLCContainerImpl> WSLCContainerImpl::Create(
...
-                createdPort.Family = AF_INET;
-                createdPort.ContainerPort = port;
-                createdPort.Protocol = protocol;
-                strcpy_s(createdPort.BindingAddress, "127.0.0.1");
+                // Exposed ports carry only a port and protocol (tcp/udp), never an address family.
+                // Mirror Docker's dual-stack default by publishing each exposed port on both the IPv4
+                // and IPv6 loopback, while keeping wslc's loopback-only default binding convention.
+                for (const auto& [family, address] : {std::pair{AF_INET, "127.0.0.1"}, std::pair{AF_INET6, "::1"}})
+                {
+                    auto& createdPort = ports.emplace_back();
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslcsession/WSLCContainer.cpp:1852]: I don't think this is valid; I believe this conflates two separate relay paths.

Rather than explain code paths, I'm updating the tests to functionally verify that not only are the ports published but they actually work for each type. That should lay to rest any sort of "this won't work" concerns from static code analysis when the test proves it correct and functional end-to-end.

File: test/windows/wslc/e2e/WSLCE2EContainerCreateTests.cpp [hunk 5]
```diff
@@ -1379,6 +1379,124 @@ class WSLCE2EContainerCreateTests
...
+udp.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
+udp.bind(('::', 9090))
+
+threading.Thread(target=http_server.serve_forever, daemon=True).start()
+print('SERVERS READY', flush=True)
+
+while True:
+    data, address = udp.recvfrom(1024)
+    udp.sendto(data.upper(), address)
+)PY");
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply test/windows/wslc/e2e/WSLCE2EContainerCreateTests.cpp:1432]: This is bogus feedback for two reasons:
1) the test passes, so it definitely successfully does the publishing
2) there are no leading spaces; the string literal deliberately starts flush to column 0 so it is valid python and easier to read the script than a chain of quoted strings

File: src/windows/wslcsession/WSLCContainer.cpp [hunk 6]
```diff
@@ -1820,22 +1823,39 @@ std::shared_ptr<WSLCContainerImpl> WSLCContainerImpl::Create(
...
                 }
 
-                auto& createdPort = ports.emplace_back();
-                createdPort.HostPort = WSLC_EPHEMERAL_PORT;
-                createdPort.Family = AF_INET;
-                createdPort.ContainerPort = port;
-                createdPort.Protocol = protocol;
-                strcpy_s(createdPort.BindingAddress, "127.0.0.1");
+                // Exposed ports carry only a port and protocol (tcp/udp), never an address family.
+                // Mirror Docker's dual-stack default by publishing each exposed port on both the IPv4
```
Comment (OneBlue) [inline src/windows/wslcsession/WSLCContainer.cpp:1848]: Looking through the tests, I'm assuming that docker is doing the ipv4 <-> ipv6 translation for us ? If so, that's pretty cool !

**Review by OneBlue — APPROVED** [review]
LGTM, thank you !

File: src/windows/wslcsession/WSLCContainer.cpp (hunk 6 above)
Comment (dkbennett) [reply src/windows/wslcsession/WSLCContainer.cpp:1848]: Yeah it just works! I was surprised at how easy it was, and the E2E test verifies the functionality.
